File: strate_ai/ai_explainer.py
"""
STRate AI — AI Explanation Engine
Uses OpenAI to convert pricing numbers → human explanation.
Role: EXPLAIN only. Never calculate.
"""
import os
from typing import Optional

try:
    from openai import OpenAI
    _OPENAI_AVAILABLE = True
except ImportError:
    _OPENAI_AVAILABLE = False

from .pricing_engine import PricingOutput, demand_label


SYSTEM_PROMPT = """You are a revenue manager for short-term rental (STR) properties in Malaysia.
Your job is to explain pricing decisions in simple, clear English to property owners.

Rules:
- Maximum 2 sentences
- Be specific: mention the price, demand level, and top reason
- Use Malaysian context (RM currency, local events)
- Don't use jargon. Speak like a helpful advisor.
- Never say "I recommend" — state it as fact.

Example outputs:
"Set RM320 tonight — strong weekend demand with 85% area occupancy makes this the optimal price."
"Drop to RM185 for tomorrow — low demand period with competitors averaging RM170 means pricing lower fills the calendar."
"""

USER_PROMPT_TEMPLATE = """Generate a pricing explanation for this property:

- Property base price: RM{base_price}
- Suggested price: RM{suggested_price}
- Price change: {change_pct:+.1f}%
- Occupancy rate in area: {occupancy_pct:.0f}%
- Competitor average price: RM{competitor_avg}
- Event nearby: {event_status}
- Weekend: {weekend_status}
- Demand level: {demand_level}
- Confidence score: {confidence:.0f}%

Breakdown of price adjustments:
- Demand factor: {demand_factor:+.1f}%
- Event boost: {event_boost:+.1f}%
- Competitor gap: {competitor_gap:+.1f}%
- Occupancy adjustment: {occupancy_adj:+.1f}%

Write 1-2 sentences explaining this recommendation to the property owner."""


def _fallback_explanation(output: PricingOutput, market_data: dict) -> str:
    """
    Rule-based fallback when OpenAI is unavailable.
    Covers the most common scenarios deterministically.
    """
    price = output.suggested_price
    base = output.breakdown["base_price"]
    change_pct = output.total_adjustment_pct
    occ = output.breakdown["occupancy_rate"]
    event = output.breakdown["event_flag"]
    is_weekend = output.breakdown["is_weekend"]
    direction = "Increase" if change_pct > 0 else "Reduce"
    action = "to" if change_pct > 0 else "to"

    reasons = []
    if event:
        reasons.append("nearby event driving demand")
    if is_weekend:
        reasons.append("weekend demand surge")
    if occ >= 0.80:
        reasons.append(f"high area occupancy ({occ*100:.0f}%)")
    elif occ <= 0.40:
        reasons.append(f"low demand period (occupancy {occ*100:.0f}%)")
    if abs(output.competitor_gap) >= 0.10:
        comp_avg = output.breakdown.get("competitor_avg", 0)
        if output.competitor_gap > 0:
            reasons.append(f"competitors averaging RM{comp_avg:.0f}")
        else:
            reasons.append(f"competitors pricing lower at RM{comp_avg:.0f}")

    reason_str = ", ".join(reasons) if reasons else "current market conditions"

    if abs(change_pct) < 1.0:
        return (
            f"Maintain RM{price:.0f} — market conditions are stable with no significant "
            f"demand shifts. Confidence: {output.confidence_score*100:.0f}%."
        )

    return (
        f"{direction} {action} RM{price:.0f} ({change_pct:+.1f}%) due to {reason_str}. "
        f"This maximizes revenue at {output.confidence_score*100:.0f}% confidence."
    )


def generate_explanation(
    output: PricingOutput,
    market_data: dict,
    api_key: Optional[str] = None,
) -> str:
    """
    Generate a human-readable explanation for the pricing recommendation.
    Falls back to rule-based text if OpenAI is unavailable.
    """
    if api_key is None:
        api_key = os.environ.get("OPENAI_API_KEY", "")

    if not _OPENAI_AVAILABLE or not api_key:
        return _fallback_explanation(output, market_data)

    base_price = output.breakdown["base_price"]
    suggested = output.suggested_price
    change_pct = ((suggested - base_price) / base_price) * 100

    prompt = USER_PROMPT_TEMPLATE.format(
        base_price=base_price,
        suggested_price=suggested,
        change_pct=change_pct,
        occupancy_pct=output.breakdown["occupancy_rate"] * 100,
        competitor_avg=output.breakdown.get("competitor_avg", "N/A"),
        event_status="Yes" if output.breakdown["event_flag"] else "No",
        weekend_status="Yes" if output.breakdown["is_weekend"] else "No",
        demand_level=demand_label(output.breakdown["occupancy_rate"], output.breakdown["event_flag"]),
        confidence=output.confidence_score * 100,
        demand_factor=output.demand_factor * 100,
        event_boost=output.event_boost * 100,
        competitor_gap=output.competitor_gap * 100,
        occupancy_adj=output.occupancy_adj * 100,
    )

    try:
        client = OpenAI(api_key=api_key)
        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
            max_tokens=120,
            temperature=0.4,
        )
        return response.choices[0].message.content.strip()
    except Exception as e:
        # Always fall back gracefully — pricing must never fail due to AI
        return _fallback_explanation(output, market_data)


def generate_whatsapp_reply(property_name: str, rec: dict) -> str:
    """
    Short WhatsApp-style reply format.
    Used by the messaging integration layer.
    """
    price = rec.get("suggested_price", 0)
    base = rec.get("base_price", price)
    change_pct = ((price - base) / base * 100) if base > 0 else 0
    reason = rec.get("reason", "")

    # Extract first sentence only for WhatsApp brevity
    short_reason = reason.split(".")[0] if reason else "current market conditions"

    direction = "▲" if change_pct >= 0 else "▼"
    return (
        f"*STRate AI — {property_name}*\n\n"
        f"💰 Set price: *RM{price:.0f}* {direction} ({change_pct:+.1f}%)\n\n"
        f"📊 {short_reason}.\n\n"
        f"_Confidence: {rec.get('confidence_score', 0)*100:.0f}%_"
    )

File: src/AiExplainer.php
<?php
/**
 * STRate AI — AI Explanation Engine
 * Calls OpenAI to convert pricing numbers → human explanation.
 * Role: EXPLAIN only. Never calculate.
 * Falls back to rule-based text if API unavailable.
 */
class AiExplainer
{
    private const SYSTEM_PROMPT = <<<PROMPT
You are a revenue manager for short-term rental (STR) properties in Malaysia.
Your job is to explain pricing decisions in simple, clear English to property owners.

Rules:
- Maximum 2 sentences
- Be specific: mention the price, demand level, and top reason
- Use Malaysian context (RM currency, local events)
- Don't use jargon. Speak like a helpful advisor.
- Never say "I recommend" — state it as fact.

Example outputs:
"Set RM320 tonight — strong weekend demand with 85% area occupancy makes this the optimal price."
"Drop to RM185 for tomorrow — low demand period with competitors averaging RM170 means pricing lower fills the calendar."
PROMPT;

    /**
     * Generate explanation via OpenAI API.
     * Falls back gracefully if API is down or key is missing.
     */
    public static function explain(array $pricingResult, array $marketData): string
    {
        if (!defined('OPENAI_API_KEY') || OPENAI_API_KEY === '') {
            return self::fallback($pricingResult, $marketData);
        }

        $bd          = $pricingResult['breakdown'];
        $suggested   = $pricingResult['suggested_price'];
        $base        = $bd['base_price'];
        $changePct   = round((($suggested - $base) / $base) * 100, 1);
        $occ         = $bd['occupancy_rate'];
        $compAvg     = $bd['competitor_avg'];
        $event       = $bd['event_flag']  ? 'Yes' : 'No';
        $weekend     = $bd['is_weekend']  ? 'Yes' : 'No';
        $confidence  = round($pricingResult['confidence_score'] * 100);
        $demandLevel = PricingEngine::demandLabel($occ, $bd['event_flag']);

        $userPrompt = <<<PROMPT
Generate a pricing explanation for this property:

- Property base price: RM{$base}
- Suggested price: RM{$suggested}
- Price change: {$changePct}%
- Occupancy rate in area: {$occ}
- Competitor average price: RM{$compAvg}
- Event nearby: {$event}
- Weekend: {$weekend}
- Demand level: {$demandLevel}
- Confidence score: {$confidence}%

Breakdown of price adjustments:
- Demand factor: {$bd['demand_factor_pct']}%
- Event boost: {$bd['event_boost_pct']}%
- Competitor gap: {$bd['competitor_gap_pct']}%
- Occupancy adjustment: {$bd['occupancy_adj_pct']}%

Write 1-2 sentences explaining this recommendation to the property owner.
PROMPT;

        try {
            $payload = json_encode([
                'model'       => OPENAI_MODEL,
                'messages'    => [
                    ['role' => 'system', 'content' => self::SYSTEM_PROMPT],
                    ['role' => 'user',   'content' => $userPrompt],
                ],
                'max_tokens'  => 120,
                'temperature' => 0.4,
            ]);

            $ch = curl_init('https://api.openai.com/v1/chat/completions');
            curl_setopt_array($ch, [
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_POST           => true,
                CURLOPT_POSTFIELDS     => $payload,
                CURLOPT_TIMEOUT        => 15,
                CURLOPT_HTTPHEADER     => [
                    'Content-Type: application/json',
                    'Authorization: Bearer ' . OPENAI_API_KEY,
                ],
            ]);

            $response = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);

            if ($httpCode === 200 && $response) {
                $data = json_decode($response, true);
                $content = $data['choices'][0]['message']['content'] ?? '';
                if ($content) return trim($content);
            }
        } catch (Throwable $e) {
            // Fall through to fallback — pricing must never fail due to AI
        }

        return self::fallback($pricingResult, $marketData);
    }

    /**
     * Rule-based fallback — always works, no API needed.
     */
    public static function fallback(array $result, array $market): string
    {
        $bd        = $result['breakdown'];
        $suggested = $result['suggested_price'];
        $base      = $bd['base_price'];
        $changePct = round((($suggested - $base) / $base) * 100, 1);
        $occ       = $bd['occupancy_rate'];
        $event     = $bd['event_flag'];
        $weekend   = $bd['is_weekend'];
        $compAvg   = $bd['competitor_avg'];
        $conf      = round($result['confidence_score'] * 100);
        $direction = $changePct >= 0 ? 'Increase' : 'Reduce';

        $reasons = [];
        if ($event)                          $reasons[] = 'nearby event driving demand';
        if ($weekend)                        $reasons[] = 'weekend demand surge';
        if ($occ >= 0.80)                    $reasons[] = sprintf('high area occupancy (%d%%)', $occ * 100);
        elseif ($occ <= 0.40)                $reasons[] = sprintf('low demand period (occupancy %d%%)', $occ * 100);
        if ($result['competitor_gap'] > 0.10)  $reasons[] = sprintf('competitors averaging RM%d', $compAvg);
        elseif ($result['competitor_gap'] < -0.10) $reasons[] = sprintf('competitors pricing lower at RM%d', $compAvg);

        $reasonStr = !empty($reasons) ? implode(', ', $reasons) : 'current market conditions';

        if (abs($changePct) < 1.0) {
            return sprintf(
                'Maintain RM%d — market conditions are stable with no significant demand shifts. Confidence: %d%%.',
                $suggested, $conf
            );
        }

        return sprintf(
            '%s to RM%d (%+.1f%%) due to %s. This maximizes revenue at %d%% confidence.',
            $direction, $suggested, $changePct, $reasonStr, $conf
        );
    }

    /**
     * Short WhatsApp-style reply for messaging integration.
     */
    public static function whatsappReply(string $propertyName, array $rec): string
    {
        $price     = (float) $rec['suggested_price'];
        $base      = (float) $rec['base_price'];
        $changePct = $base > 0 ? round((($price - $base) / $base) * 100, 1) : 0;
        $arrow     = $changePct >= 0 ? '▲' : '▼';
        $reason    = $rec['reason'] ?? '';

        // First sentence only for brevity
        $shortReason = strstr($reason, '.', true) ?: $reason;

        return sprintf(
            "*STRate AI — %s*\n\n💰 Set price: *RM%d* %s (%+.1f%%)\n\n📊 %s.\n\n_Confidence: %d%%_",
            $propertyName,
            $price,
            $arrow,
            $changePct,
            $shortReason,
            round((float)$rec['confidence_score'] * 100)
        );
    }
}
